### ISSUE TITLE
Broken links on single user servers if "@" symbol in username

### ISSUE THREAD

**Issue by CedricFch** [jupyterhub/jupyterhub#345]
If a username contains '@' symbol, some bugs appear in single-user servers because the '@' symbol is escaped as '%40' for API links.

I solved this issue except a few links (for example when stopping the server) by putting at line 118 of User.py:
base_url=url_path_join(base_url, 'user', self.escaped_name.replace("@",".")),

 but I don't know whether this is the best way to deal with this issue...

**Comment by minrk** [comment]
This is a bug in the single-user notebook package, and should be fixed by the upcoming release of notebook-4.1.

**Comment by yuvipanda** [comment]
I'm running into the same issue I believe - can you point out the commit responsible for fixing it so I can cherry-pick it for my production install? Thanks

**Comment by minrk** [comment]
I think it's a few fixes, but the primary PR is jupyter/notebook#472. Can you check if it works using jupyterhub and notebook master?

**Comment by yuvipanda** [comment]
I just tested this on notebook 4.1b1 and didn't fix this :(

**Comment by minrk** [comment]
It's possible that the way it's fixed in the notebook requires another change on the Hub's side.  I'll look into it. I know I've used it.

**Comment by minrk** [comment]
I just tested in a fresh env with

    $ pip install --pre notebook jupyterhub
    $ pip list
    appnope (0.1.0)
    decorator (4.0.6)
    gnureadline (6.3.3)
    ipykernel (4.2.1)
    ipython (4.0.1)
    ipython-genutils (0.1.0)
    Jinja2 (2.8)
    jsonschema (2.5.1)
    jupyter-client (4.1.1)
    jupyter-core (4.0.6)
    jupyterhub (0.3.0)
    MarkupSafe (0.23)
    mistune (0.7.1)
    nbconvert (4.1.0)
    nbformat (4.0.1)
    notebook (4.1.0b1)
    pamela (0.2.1)
    path.py (8.1.2)
    pexpect (4.0.1)
    pickleshare (0.5)
    pip (7.1.2)
    ptyprocess (0.5)
    Pygments (2.0.2)
    pyzmq (15.1.0)
    requests (2.8.1)
    setuptools (18.2)
    simplegeneric (0.8.1)
    SQLAlchemy (1.0.10)
    terminado (0.5)
    tornado (4.3)
    traitlets (4.1.0b1)
    wheel (0.24.0)

and it works with `@` in the username and URLs. What do you see as not working?

**Comment by yuvipanda** [comment]
Hmm, I checked with '(' in the username. Let me try again

**Comment by yuvipanda** [comment]
I have:

```
YuviPanda_(WMF)@PAWS:~$ pip freeze
backports-abc==0.4
bash-kernel==0.4.1
beautifulsoup4==4.4.1
bokeh==0.10.0
codecov==1.6.3
coverage==4.0.3
crontab==0.20.3
cycler==0.9.0
decorator==4.0.6
docopt==0.6.2
flake8==2.5.1
Flask==0.10.1
flickrapi==2.1.2
google==1.7
inflect==0.2.5
ipykernel==4.2.1
ipython==4.0.1
ipython-genutils==0.1.0
ipywidgets==4.1.1
irc==13.2
itsdangerous==0.24
jaraco.apt==1.0
jaraco.classes==1.2
jaraco.collections==1.3.1
jaraco.context==1.5.2
jaraco.functools==1.8.1
jaraco.itertools==1.7
jaraco.logging==1.2
jaraco.text==1.4
jaraco.timing==1.1.5
Jinja2==2.8
jsonable==0.3.1
jsonschema==2.5.1
jupyter-client==4.1.1
jupyter-core==4.0.6
jupyterhub==0.3.0
MarkupSafe==0.23
matplotlib==1.5.0
mccabe==0.3.1
memento-client==0.5.3
mistune==0.7.1
more-itertools==2.2
mwcli==0.0.2
mwdb==0.1.0
mwoauth==0.2.4
mwparserfromhell==0.4.3
mwtypes==0.2.0
mwxml==0.2.1
nbconvert==4.1.0
nbformat==4.0.1
nose==1.3.7
notebook==4.1.0rc1
numpy==1.10.2
oauthlib==1.0.3
pamela==0.2.1
pandas==0.17.1
para==0.0.5
path.py==8.1.2
pep257==0.7.0
pep8==1.5.7
pexpect==4.0.1
pickleshare==0.5
ptyprocess==0.5
pycountry==1.18
pydot==1.0.29
pyflakes==1.0.0
Pygments==2.0.2
PyJWT==1.4.0
PyMySQL==0.6.7
pyparsing==2.0.6
python-dateutil==2.4.2
python-mwapi==0.2
python-stdnum==1.2
pytz==2015.7
pywikibot==2.0rc1.post2
PyYAML==3.11
pyzmq==15.1.0
requests==2.9.1
requests-oauthlib==0.6.0
requests-toolbelt==0.5.1
simplegeneric==0.8.1
six==1.10.0
socketIO-client==0.5.6
SQLAlchemy==1.0.11
tempora==1.3
terminado==0.5
tornado==4.3
traitlets==4.1.0b1
websocket-client==0.34.0
Werkzeug==0.11.2
yg.lockfile==2.1
zc.lockfile==1.1.0
YuviPanda_(WMF)@PAWS:~$ 
```

And when I try to https://tools.wmflabs.org/paws/user/YuviPanda_(WMF)/ (which is where I'm redirected to) I get a 404. If I hand-change the URL to https://tools.wmflabs.org/paws/user/YuviPanda_%28WMF%29/tree# I get an empty file browser, and all actions fail (opening terminal has no terminal, etc). This was the same behavior as 4.0.x as well.

**Comment by willingc** [comment]
@minrk Perhaps a shot in the dark, but I wonder if the ``hub_api.url`` needs to be escaped before the ``join`` on L68 in this file: https://github.com/jupyter/jupyterhub/blob/master/jupyterhub/singleuser.py#L66

**Comment by yuvipanda** [comment]
This still happens in user names with ( in them, in the 4.1 release.

**Comment by tmtabor** [comment]
I've been trying to track down this very issue, and here's my understanding of what's going wrong:

In user.py (on line 149 as of version 4.1) the User object is escaping special characters in usernames, but explicitly not encoding @ symbols:
`return quote(self.name, safe='@')`

This escaped username is being used for both cookies and to generate URLs. For example, User.base_url is being set in user.py on line 112, using the escaped name, without @'s being escaped.

The problem, as far as I can tell, is that the code on the notebook side expects %40 in the URL rather than the @ symbol, and that results in some calls returning 404's, as JupyterHub is proxying `/user/foo@bar` rather than `/user/foo%40bar`.

**Comment by minrk** [comment]
@tmtabor that's fixed in notebook-4.1. What version of the notebook package are you using?

**Comment by mariusvniekerk** [comment]
I'm also seeing this on my jupyterhub deploy.  I'm using an older version of the hub and notebook-4.1.  Do I need to upgrade the hub?

**Comment by minrk** [comment]
@mariusvniekerk likely. Let me know if you can reproduce it with notebook-4.1 and jupyterhub-0.4

**Comment by tmtabor** [comment]
I am running notebook-4.1.0 and jupyterhub-0.4.1 (using a variant of OAuthenticator and DockerSpawner). I am still seeing issues when logging in as a user with a @ in the username.

A lot of JavaScript imports on the Jupyter-side are breaking, which is in turn breaking functionality in the notebook. Here are a couple of the errors I am seeing in my JavaScript console:

> GET http://gpnotebook.broadinstitute.org/user/tabor%40broadinstitute.org/api/config/tree?_=1455120998006 404 Not Found 8ms	
> GET http://gpnotebook.broadinstitute.org/user/tabor%4...dinstitute.org/api/config/common?_=1455120998007 404 Not Found 28ms	
> GET http://gpnotebook.broadinstitute.org/user/tabor%40broadinstitute.org/api/terminals?_=1455120998008 404 Not Found 28ms	
main.mi...81acc18 (line 6)

The errors go on and on like that. But notice how the notebook is trying to load the files with the @ symbol encoded as %40, while JupyerHub doesn't encode the @ symbol.

**Comment by minrk** [comment]
Yeah, that's exactly the symptom of the bug. What's weird is that I just checked and verified that my server is indeed working correctly with jupyterhub-0.4 and notebook-4.1 and @ in usernames. It doesn't work with 4.0, but upgrading notebook to 4.1 fixes the bug, as expected.

What do you get from `https://yourhub/user/you@email.com/api`?

**Comment by tmtabor** [comment]
Actually, you can disregard most of what I just said. It looks like when I updated the notebook from 4.0 to 4.1 inside the Docker container, I did so in the wrong Python environment (pip vs conda).

Having actually updated, however, I am still seeing a few similar errors in the browser console, although vastly fewer than before. Here's a selection from the console:

> GET http://gpnotebook.broadinstitute.org/user/tabor@broadinstitute.org/api/sessions?_=1455125865768 200 OK 10ms
> GET http://gpnotebook.broadinstitute.org/user/tabor@broadinstitute.org/api/terminals?_=1455125865769 200 OK 17ms
> GET http://gpnotebook.broadinstitute.org/user/tabor%40broadinstitute.org/api/contents?type=directory&_=1455125865770 404 Not Found 7ms

**Comment by anderbubble** [comment]
We just upgraded to 0.4.1 on recommendation from @minrk, and our issue still persists. See errors like this in singleuser-server:

    [W 2016-02-26 11:13:43.064 ${user}@${domain} log:47] 404 GET /user/${user}%40${domain}/api/config/tree?_=1456510428994 (128.138.138.146) 48.72ms referer=https://jupyter1.${domain}/user/${user}@${domain}/tree
    [D 2016-02-26 11:13:43.066 ${user}@${domain} handlers:166] Using contents: services/contents
    [W 2016-02-26 11:13:43.069 ${user}@${domain} log:47] 404 GET /user/${user}%40${domain}/api/config/common?_=1456510428995 (128.138.138.146) 3.90ms referer=https://jupyter1.${domain}/user/${user}@${domain}/tree
    [D 2016-02-26 11:13:43.071 ${user}@${domain} handlers:166] Using contents: services/contents
    [W 2016-02-26 11:13:43.074 ${user}@${domain} log:47] 404 GET /user/${user}%40${domain}/api/terminals?_=1456510428996 (128.138.138.146) 3.67ms referer=https://jupyter1.${domain}/user/${user}@${domain}/tree
    [D 2016-02-26 11:13:43.076 ${user}@${domain} handlers:166] Using contents: services/contents
    [W 2016-02-26 11:13:43.078 ${user}@${domain} log:47] 404 GET /user/${user}%40${domain}/api/kernelspecs (128.138.138.146) 3.74ms referer=https://jupyter1.${domain}/user/${user}%40${domain}/tree
    [D 2016-02-26 11:13:43.081 ${user}@${domain} handlers:166] Using contents: services/contents
    [W 2016-02-26 11:13:43.084 ${user}@${domain} log:47] 404 GET /user/${user}%40${domain}/api/sessions?_=1456510428997 (128.138.138.146) 4.15ms referer=https://jupyter1.${domain}/user/${user}@${domain}/tree
    [D 2016-02-26 11:13:43.086 ${user}@${domain} handlers:166] Using contents: services/contents
    [W 2016-02-26 11:13:43.089 ${user}@${domain} log:47] 404 GET /user/${user}%40${domain}/api/terminals?_=1456510428998 (128.138.138.146) 3.77ms referer=https://jupyter1.${domain}/user/${user}@${domain}/tree
    [D 2016-02-26 11:14:06.399 ${user}@${domain} handlers:166] Using contents: services/contents
    [W 2016-02-26 11:14:06.403 ${user}@${domain} log:47] 404 GET /user/${user}%40${domain}/api/sessions?_=1456510428999 (128.138.138.146) 4.55ms referer=https://jupyter1.${domain}/user/${user}@${domain}/tree
    [D 2016-02-26 11:14:06.404 ${user}@${domain} handlers:166] Using contents: services/contents
    [W 2016-02-26 11:14:06.405 ${user}@${domain} log:47] 404 GET /user/${user}%40${domain}/api/terminals?_=1456510429000 (128.138.138.146) 2.24ms referer=https://jupyter1.${domain}/user/${user}@${domain}/tree

**Comment by anderbubble** [comment]
I've just now understood that this is supposed to be a bug in notebook <4.1, not jupyterhub; so I'll update that and try again. (We're on 4.0.2)

**Comment by yuvipanda** [comment]
We did another round of fixups related to URL encoding in JupyterHub a few weeks ago, and my issues have been fixed.

**Comment by willingc** [comment]
Closing this issue since it appears to be resolved based on @yuvipanda's report. Please feel free to leave an issue at [jupyter/help](https://github.com/jupyter/help) or leave a comment requesting this issue be reopened. Thanks to all!
